File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int maxLen = Integer.MIN_VALUE, left = 0;
        Set<Character> set = new HashSet<>(); 

        for(int right = 0; right < s.length(); right++){
            
            while(set.contains(s.charAt(right))){
                set.remove(s.charAt(left));
                left++;
            }

            set.add(s.charAt(right));
            maxLen = Math.max(maxLen, right - left + 1);
        }

        return maxLen == Integer.MIN_VALUE ? 0 : maxLen;
    }  
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        mp = collections.defaultdict(int)
        ans=0
        j=0
        n = len(s)
        for i in range(0, n):
            c = s[i]
            mp[c]+=1
            while mp[c]>1:
                mp[s[j]]-=1
                j+=1
            ans=max(ans, i-j+1)
        return ans